### ISSUE TITLE
Binary data compare failed on multi-volume archive with stored mode and QO information.

### ISSUE THREAD

**Issue by jyhpsycho** [hasse69/rar2fs#44]
* Originally it's [comment 14](https://github.com/hasse69/rar2fs/issues/42#issuecomment-165829091) of issue #42.

Instructions for preparing provided sample:
```
mkdir TEST_DIR O
mkdir TEST_DIR/DIR{1..3}
mkdir TEST_DIR/DIR1/{A,B}
mkdir TEST_DIR/DIR2/C
dd if=/dev/urandom of=TEST_DIR/DIR1/A/FILE1 bs=4k count=16
dd if=/dev/urandom of=TEST_DIR/DIR1/A/FILE2 bs=4k count=16
dd if=/dev/urandom of=TEST_DIR/DIR1/B/FILE3 bs=4k count=16
dd if=/dev/urandom of=TEST_DIR/DIR2/C/FILE4 bs=4k count=16
dd if=/dev/urandom of=TEST_DIR/DIR2/FILE5 bs=4k count=16
rar a -m0 -ma5 -v16000b -rr5 -qo+ TEST.rar TEST_DIR
./rar2fs-master/rar2fs -o ro,allow_other --flat-only --show-comp-img TEST.part01.rar O
diff -ur TEST_DIR O/TEST_DIR
```

Note that it uses -qo+ option. It stores 'Quick open information' for all files in RAR archive. With -qo- option, it doesn't occur.
Above instruction uses -rr5 option, but recovery record is not related on this issue. Same result without it.

NOTE : Directory hierarchy building provided sample is created on tmpfs; It returns the file list as reverse order of being created. It's not important thing for this issue, but one may confuse that it does not store files as typical order.

[TEST.zip](https://github.com/hasse69/rar2fs/files/66866/TEST.zip)

**Comment by hasse69** [comment]
Thanks for the report. I will however tag this as enhancement for now.
"Quick Open" is a RAR5 specific feature. It is currently not supported and I am not sure I will ever get the time to actually implement it either. What is needed to to make it work transparently in rar2fs and more or less ignore it. There is not point in actually implementing the "Quick Open" scheme.

**Comment by jyhpsycho** [comment]
Right. It's just introduced for 'open archove quickly' as its name, combine all file header in archive to near end of archive. It's just copy of individual file header, thus it can be skipped safely.

**Comment by jyhpsycho** [comment]
I found that rar2fs reads much less data than stored.

```
UNRAR 5.30 freeware      Copyright (c) 1993-2015 Alexander Roshal

Archive: TEST.part19.rar
Details: RAR 5, volume 19, recovery record

        Name: TEST_DIR/DIR1/A/FILE1
        Type: File
        Size: 65536
 Packed size: 14610
       Ratio: <->
       mtime: 2015-12-19 01:25:54,973
  Attributes: ..A....
  Pack-CRC32: CA0AB508
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: QO
        Type: Service
        Size: 65
 Packed size: 65
       Ratio: 100%
  Attributes: .B
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: RR
        Type: Service
        Size: 1170
 Packed size: 1170
       Ratio: 100%
  Attributes: .B
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

     Service: EOF
```

```
Opening /mnt/WORK/4/TEST.part18.rar
SEEK src_off = 14280, VOL_REAL_SZ = 14625
size = 16384, chunk = 345
Read 345 bytes from vol=0, base=18
get_vname()   /mnt/WORK/4/TEST.part18.rar   vol=19, len=2, pos=21
Opening /mnt/WORK/4/TEST.part19.rar
SEEK src_off = 87, VOL_REAL_SZ = 14625
size = 16039, chunk = 14538
Read 14538 bytes from vol=1, base=18
get_vname()   /mnt/WORK/4/TEST.part18.rar   vol=20, len=2, pos=21
Opening /mnt/WORK/4/TEST.part20.rar
SEEK src_off = 87, VOL_REAL_SZ = 14625
size = 1501, chunk = 14538
Read 1501 bytes from vol=2, base=18
   read[140510459685440] 16384 bytes from 0
   unique: 98, success, outsize: 16400
```

It should read 14610 bytes from 2nd volume, but just 14538. Because of that, it misses 72-bytes of data per volume, larger than QO data. It seems that it uses mis-caluclated values... Is rar2fs does not use header information with RAW I/O mode?

**Comment by jyhpsycho** [comment]
I found very strange behavior. It does not occur any error when comparing individual files, e.g. `diff -ur O/TEST_DIR/DIR1/A/FILE1 TEST_DIR/DIR1/A/FILE1`. The error only occurs when running `diff -ur O/TEST_DIR TEST_DIR`.

**Comment by jyhpsycho** [comment]
`diff -ur O/TEST_DIR/DIR1/A/FILE1 TEST_DIR/DIR1/A/FILE1` using lread(), not lread_raw(). I guess that some login on lread_raw() is broken.

**Comment by jyhpsycho** [comment]
VOL_NEXT_SZ used by lread_raw() has incorrect value on that case. It's assigned by list_rar().

**Comment by jyhpsycho** [comment]
It assumes that splitted file data is ended at same location as 1st volume contains such file. But, it can vary due to QO data.
```
UNRAR 5.30 freeware      Copyright (c) 1993-2015 Alexander Roshal

Archive: TEST.part05.rar
Details: RAR 5, volume 5, recovery record

        Name: TEST_DIR/DIR2/FILE5
        Type: File
        Size: 65536
 Packed size: 7079
       Ratio: <--
       mtime: 2015-12-19 01:26:15,482
  Attributes: ..A....
       CRC32: AFC59CBA
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: TEST_DIR/DIR2/C/FILE4
        Type: File
        Size: 65536
 Packed size: 7407
       Ratio: -->
       mtime: 2015-12-19 01:26:11,243
  Attributes: ..A....
  Pack-CRC32: 037937D4
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: QO
        Type: Service
        Size: 128
 Packed size: 128
       Ratio: 100%
  Attributes: .B
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: RR
        Type: Service
        Size: 1170
 Packed size: 1170
       Ratio: 100%
  Attributes: .B
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

     Service: EOF
```
```
UNRAR 5.30 freeware      Copyright (c) 1993-2015 Alexander Roshal

Archive: TEST.part06.rar
Details: RAR 5, volume 6, recovery record

        Name: TEST_DIR/DIR2/C/FILE4
        Type: File
        Size: 65536
 Packed size: 14610
       Ratio: <->
       mtime: 2015-12-19 01:26:11,243
  Attributes: ..A....
  Pack-CRC32: 9D542247
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: QO
        Type: Service
        Size: 65
 Packed size: 65
       Ratio: 100%
  Attributes: .B
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

        Name: RR
        Type: Service
        Size: 1170
 Packed size: 1170
       Ratio: 100%
  Attributes: .B
     Host OS: Windows
 Compression: RAR 5.0(v50) -m0 -md=128K

     Service: EOF
```

For example, the end point of file data is different at least 63(128 - 65) bytes on case above. It causes missing some data when reading. It's not lread_raw() issue, right?

**Comment by jyhpsycho** [comment]
For fixing this issue 'transparently', just comment out line [#2387](https://github.com/hasse69/rar2fs/blob/master/rar2fs.c#L2387) of rar2fs.c. It just disables raw access, and it works correctly as you also commented at [here](https://github.com/hasse69/rar2fs/issues/42#issuecomment-165895569).
```
entry_p->flags.raw = 1;
```

Implementing raw access mode considering QO seems to not easy... It cannot be simply calculated by using 1st volume's information only. QO can contain any combination of File Header and Service Header. There's no restriction that it should have all of them(from [Technote](http://www.rarlab.com/technote.htm)), although specifying -qo+ option enforces it. By default, WinRAR uses own logic to determine whether the header should(?) be included in QO or not.

Most flexible way to do it is, "Read all volume contains splitted file, and store individual data block size and use it". It will be work even every volume has different size, normally it can't be. The second way is, "Read next(2nd) volume data and set VOL_NEXT_SZ value using that". It will be work if all volumes except last has same size.

**Comment by hasse69** [comment]
Yes, I need to check how this was handled for RAR4 archives.
I need to look at some solution to prefetch information from at least the first and the second volume file perhaps.

**Comment by hasse69** [comment]
Tested temporary fix for issue #42 and it seems to resolve also this issue.
I will close this one as a duplicate since issue #42 covers a wider use-case.
